libarchive/archive_write_set_format_7zip.c

```diff
@@ -420,16 +452,47 @@ _7z_options(struct archive_write *a, const char *key, const char *value)
 		return (ARCHIVE_OK);
 	}
 	if (strcmp(key, "compression-level") == 0) {
-		if (value == NULL ||
-		    !(value[0] >= '0' && value[0] <= '9') ||
-		    value[1] != '\0') {
-			archive_set_error(&(a->archive),
-			    ARCHIVE_ERRNO_MISC,
-			    "Illegal value `%s'",
-			    value);
+		if (value == NULL || *value == '\0') {
+			archive_set_error(&(a->archive), ARCHIVE_ERRNO_MISC,
+				"Invalid compression-level option value `%s'", value);
+			return (ARCHIVE_FAILED);
+		}
+
+		errno = 0;
+		char *end = NULL;
+		int lvl = (int)strtol(value, &end, 10);
+
+		if (errno) {
+			archive_set_error(&(a->archive), errno,
+				"parsing compression-level option value failed `%s'", value);
+			return (ARCHIVE_FAILED);
+		}
+		if (end == NULL || *end != '\0') {
+			archive_set_error(&(a->archive), ARCHIVE_ERRNO_MISC,
+				"parsing compression-level option value failed `%s'", value);
 			return (ARCHIVE_FAILED);
 		}
-		zip->opt_compression_level = value[0] - '0';
+
+#if HAVE_ZSTD_H
+		if (lvl < ZSTD_minCLevel() || lvl > ZSTD_maxCLevel()) {
```

Comment:
`ZSTD_minCLevel()` is only available in zstd >= v1.4.0, I guess I should add an autoconf check before using it.